balancer-js/src/modules/pools/pool-types/concerns/composableStable/exit.concern.ts

```diff
@@ -38,16 +153,361 @@ export class ComposableStablePoolExit implements ExitConcern {
     slippage,
     wrappedNativeAsset,
   }: ExitExactTokensOutParameters): ExitExactTokensOutAttributes => {
-    // TODO implementation
-    console.log(
-      exiter,
+    this.checkInputsExactTokensOut(tokensOut, amountsOut, pool);
+    const sortedValues = this.getValuesSorted({
       pool,
-      tokensOut,
+      wrappedNativeAsset,
       amountsOut,
+      tokensOut,
+    });
+    if (!sortedValues.upScaledAmountsOut) {
+      throw Error('amountsOut or tokensOut parameter invalid;');
+    }
+    //DOING THE MATHS TO CALCULATE THE bptIn AND maxBPTIn VALUES
+    const { bptIn, maxBPTIn } = this.doMathsV1BPTInExactTokensOut({
+      ...sortedValues,
       slippage,
-      wrappedNativeAsset
+    });
+
+    const userData = ComposableStablePoolEncoder.exitBPTInForExactTokensOut(
+      sortedValues.minAmountsOutWithoutBpt as string[],
+      maxBPTIn
+    );
+
+    return this.doExitExactTokensOut({
+      ...sortedValues,
+      userData,
+      exiter,
+      poolId: pool.id,
+      bptIn,
+      maxBPTIn,
+    });
+  };
+
+  /**
+   * Checks if the input of buildExitExactBPTIn is valid;
+   */
+  checkInputsExactBptIn = (
+    bptIn: string,
+    singleTokenMaxOut: string | undefined,
+    pool: Pool,
+    shouldUnwrapNativeAsset: boolean
+  ): void => {
+    if (!bptIn.length || parseFixed(bptIn, 18).isNegative()) {
+      throw new BalancerError(BalancerErrorCode.INPUT_OUT_OF_BOUNDS);
+    }
+    if (
+      singleTokenMaxOut &&
+      singleTokenMaxOut !== AddressZero &&
+      !pool.tokens
+        .map((t) => t.address)
+        .some((a) => isSameAddress(a, singleTokenMaxOut))
+    ) {
+      throw new BalancerError(BalancerErrorCode.TOKEN_MISMATCH);
+    }
+
+    if (!shouldUnwrapNativeAsset && singleTokenMaxOut === AddressZero)
+      throw new Error(
+        'shouldUnwrapNativeAsset and singleTokenMaxOut should not have conflicting values'
+      );
+
+    //V1 COMPOSABLE STABLE POOLS ONLY SUPPORT SINGLE TOKEN OUT JOINS
+    if (!singleTokenMaxOut && pool.poolTypeVersion === 1) {
+      throw new BalancerError(BalancerErrorCode.UNSUPPORTED_POOL_TYPE_VERSION);
+    }
+
+    // Check if there's any relevant composable stable pool info missing
+    if (pool.tokens.some((token) => !token.decimals))
+      throw new BalancerError(BalancerErrorCode.MISSING_DECIMALS);
+    if (!pool.amp) throw new BalancerError(BalancerErrorCode.MISSING_AMP);
+  };
+  /**
+   * Checks if the input of buildExitExactTokensOut is valid;
+   */
+  checkInputsExactTokensOut = (
+    tokensOut: string[],
+    amountsOut: string[],
+    pool: Pool
+  ): void => {
+    if (
+      tokensOut.length != amountsOut.length ||
+      tokensOut.length != pool.tokensList.length - 1
+    ) {
+      throw new BalancerError(BalancerErrorCode.INPUT_LENGTH_MISMATCH);
+    }
+    if (
```

Comment:
This seems non-obvious to follow. At least add a comment to explain what its for.